src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py

```diff
@@ -0,0 +1,719 @@
+                      legends=legends_plot_loss,
+                      file_path=base_path_his,
+                      file_name="Loss_Branch_{}".format(branch_idx),
+                      axis_x="Training Epoch",
+                      axis_y="Loss")
+
+        # Save plots
+        np.save(path.join(base_path_his, "Loss_Branch_{}".format(branch_idx)), np.array(his_loss))
+        np.save(path.join(base_path_his, "Acc_Branch_{}".format(branch_idx)), np.array(his_acc))
+        np.save(path.join(base_path_his, "Loss_Val_Branch_{}".format(branch_idx)), np.array(his_val_loss))
+        np.save(path.join(base_path_his, "Acc_Val_Branch_{}".format(branch_idx)), np.array(his_val_acc))
+
+    def infer(self, input_batch: Union[Image, List[Image], torch.Tensor]):
+        """
+        This method is used to perform inference on an image or a batch of images.
+
+        :param input_batch: a batch of images
+        :return: dimensional and categorical emotion results.
+        """
+        if not isinstance(input_batch, (Image, list)):
+            input_batch = Image(input_batch)
+        if type(input_batch) is Image:
+            input_batch = [input_batch]
+        if type(input_batch) is list:
+            input_batch = torch.stack([torch.tensor(v.data) for v in input_batch])
+        # print('input_batch shape', input_batch.shape)
+        cpu_device = torch.device('cpu')
+
+        input_batch = input_batch.to(device=self.device, dtype=torch.float)
+        self.model.eval()
+        out_emotions, out_va = self.model(input_batch)
+
+        # categorical result
+        softmax_ = nn.Softmax(dim=0)
+        categorical_results = out_emotions[:self.ensemble_size]  # a list of 9 or (n) torch tensors
+        overall_emotion_preds = torch.zeros(categorical_results[0].size()).to(self.device)  # size: batchsize * 8
+        for o_eval, outputs_per_branch_eval in enumerate(categorical_results, 0):
+            _, preds_indices = torch.max(outputs_per_branch_eval, 1)
+            for v_i, v_p in enumerate(preds_indices, 0):
+                overall_emotion_preds[v_i, v_p] += 1
+        ensemble_emotion_results = [Category(prediction=int(o.argmax(dim=0)), confidence=softmax_(o),
+                                    description=datasets.AffectNetCategorical.get_class(int(o.argmax(dim=0))))
+                                    for o in overall_emotion_preds]
+        # dimension result
+        dimensional_results = out_va[:self.ensemble_size]
+        overall_dimension_preds = torch.zeros(dimensional_results[0].size()).to(cpu_device)
+        for evaluate_branch in range(self.ensemble_size):
+            out_va_eval_cpu = out_va[evaluate_branch].detach().to(cpu_device)
+            overall_dimension_preds += out_va_eval_cpu
+        ensemble_dimension_results = overall_dimension_preds / self.ensemble_size
+
+        return ensemble_emotion_results, ensemble_dimension_results
```

Comment:
Please check existing data types (e.g., [Category](https://github.com/opendr-eu/opendr/blob/master/src/opendr/engine/target.py#L105) ) to check if they can be re-used in order to be consistent with the way the rest of the tools work.

Reply:
Infer returns two outputs. One is ensemble_emotion_results which is categorical emotion predictions and it is of data type Category. The other is ensemble_dimension_results which is valence-arousal continuous values and it is returned just as real numbers.